### PR TITLE
feat: added datasource icon on left of entity name, revamped name editor

### LINKED ISSUE #10542: [Feature]: Indicate type of Datasource next to Query Name in Query Editor Page
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Need to add an image of Datasource type just before the query name to indicate which query the user is editing.

### Why should this be worked on?

It confuses the user about the type of datasource, one is working on.

### PR DESCRIPTION
**Pull request by AmanAgarwal041**
## Description

1. Added Datasource Icon on the left of the query name on the Query Editor page to signify the type of editing datasource query
2. There were multiple components with similar logic to identify the conflicting names when editing names of JSObject, Query Name, PageName, Names in the Entity Explorer Menu. Revamped and added a new wrapper component for Multiple Components so that logic for name editing remains at one point.

Fixes #10542 

## Type of change

> Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

1. Changing Name from the Query Editor page and see if it reflects on the entity explorer as well.
2. Changing Name from the Page Editor page and see if it reflects on the entity explorer as well.
3. Changing Name from JS Object Editor page and see if it reflects on the entity explorer as well.
4. Check if typing Conflicting Names, it shows an error or not.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/datasource-icon 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.05 **(0.04)** | 36.36 **(0.02)** | 34.7 **(0.01)** | 55.43 **(0.04)**
 :green_circle: | app/client/src/components/editorComponents/ActionNameEditor.tsx | 50 **(25.71)** | 19.05 **(15.28)** | 0 **(0)** | 50 **(24.24)**
 :sparkles: :new: | **app/client/src/components/utils/NameEditorComponent.tsx** | **18.75** | **0** | **0** | **19.57**
 :red_circle: | app/client/src/constants/messages.ts | 74.78 **(-0.05)** | 100 **(0)** | 25.24 **(-0.17)** | 80.12 **(-0.19)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/Name.tsx | 58.62 **(4.36)** | 45.45 **(13.37)** | 33.33 **(7.01)** | 57.41 **(4.08)**
 :green_circle: | app/client/src/pages/Editor/JSEditor/JSObjectNameEditor.tsx | 51.85 **(29.63)** | 20 **(15.74)** | 0 **(0)** | 51.85 **(28.12)**
 :sparkles: :new: | **app/client/src/selectors/actionSelectors.tsx** | **44.44** | **0** | **0** | **44.44**
 :red_circle: | app/client/src/selectors/ui.tsx | 83.33 **(-16.67)** | 100 **(0)** | 50 **(-50)** | 75 **(-25)**
 :red_circle: | app/client/src/selectors/widgetSelectors.ts | 62.5 **(-25)** | 35.71 **(-26.79)** | 57.14 **(-22.86)** | 61.9 **(-30.41)**
 :green_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0.78)** | 72.55 **(1.96)** | 100 **(0)** | 93.33 **(0.95)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**</details>

### CODE REVIEW

File: app/client/src/pages/Editor/JSEditor/JSObjectNameEditor.tsx [hunk 1]
```diff
@@ -54,120 +52,56 @@ type ActionNameEditorProps = {
...
+              currentJSObjectConfig ? currentJSObjectConfig.name : ""
+            }
+            editInteractionKind={NewEditInteractionKind.SINGLE}
+            fill
+            forceDefault={forceUpdate}
+            hideEditIcon
+            isEditingDefault={isNew}
+            isInvalid={isInvalidNameForEntity}
+            onBlur={handleNameChange}
+            placeholder="Name of the object in camelCase"
```
Comment (mohanarpit) [inline app/client/src/pages/Editor/JSEditor/JSObjectNameEditor.tsx:93]: Minor: Please use `messages.ts` for all string values being displayed to the user.

File: app/client/src/components/editorComponents/ActionNameEditor.tsx [hunk 2]
```diff
@@ -55,131 +59,70 @@ type ActionNameEditorProps = {
...
+        handleNameChange,
+        isInvalidNameForEntity,
+        isNew,
+        saveStatus,
+      }: {
+        forceUpdate: boolean;
+        handleNameChange: (value: string) => void;
+        isInvalidNameForEntity: (value: string) => string | boolean;
+        isNew: boolean;
+        saveStatus: any;
```
Comment (ApekshaBhosale) [inline app/client/src/components/editorComponents/ActionNameEditor.tsx:92]: add type here

**Comment by ApekshaBhosale** [comment]
@AmanAgarwal041 Did you create an issue for tab related rename bug we encountred? how are taking care of that scenario?

**Comment by AmanAgarwal041** [comment]
> @AmanAgarwal041 Did you create an issue for tab related rename bug we encountred? how are taking care of that scenario?

I haven't created the bug yet. From Property Pane the rename thing doesn't handle the conflicts but from entity explorer it does. The logic in both is totally different in both scenarios. I have revamped the current functionality right now to follow a single path for the conflict logic. I will create the issue for the bug.

File: app/client/src/components/utils/NameEditorComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import { toggleShowDeviationDialog } from "actions/onboardingActions";
+import {
+  getUsedActionNames,
+  getSavingStatusForActionName,
+} from "selectors/actionSelectors";
+
+type NameEditorProps = {
+  checkForGuidedTour?: boolean;
+  children: (params: any) => JSX.Element;
+  currentActionConfig: any;
```
[the remark this answers is not in the document]
Comment (AmanAgarwal041) [reply app/client/src/components/utils/NameEditorComponent.tsx]: fixed the type

File: app/client/src/components/editorComponents/ActionNameEditor.tsx (hunk 2 above)
Comment (AmanAgarwal041) [reply app/client/src/components/editorComponents/ActionNameEditor.tsx]: added the type

**Review by ApekshaBhosale — APPROVED** [review]

**Pull request merged**